list_test.py

```diff
@@ -0,0 +1,114 @@
+from list import List
+
+
+def create_list():
+    l = List()
+
+    l.push('p')
+    l.push('o')
+    l.push('n')
+    l.push('y')
+
+    return l
+
+
+def test_getitem_success():
+    assert create_list()[0] == 'p'
+
+def test_getitem_fail():
+    assert create_list()[0] != 'a'
+
+
+def test_setitem_success():
+    l1 = create_list()
+    l2 = create_list()
+
+    l1[-1] = 'k'
+    l2[-1] = 'k'
+
+    assert l1 == l2
+
+def test_setitem_fail():
+    l1 = create_list()
+    l2 = create_list()
+
+    l1[-1] = 'k'
+    l2[-1] = 't'
+
+    assert l1 != l2
+
+
+def test_equals_success():
+    l1 = create_list()
+    l2 = create_list()
+
+    assert l1 == l2
+
+def test_equals_fail_element():
+    l1 = create_list()
+    l1[0] = 'd'
+
+    assert l1 != create_list()
+
+def test_equals_fail_len():
+    l1 = create_list()
+    l1.remove(0)
+    
+    assert l1 != create_list()
+
+
+def test_push_success():
+    l1 = create_list()
+
+    l1.push('t')
+
+    assert len(l1) == len(create_list()) + 1 and l1[-1] == 't'
+
+def test_push_fail():
+    l1 = create_list()
+
+    l1.push('t')
+
+    assert len(l1) - 1 == len(create_list())
+
+
+def test_insert_success():
+    l1 = create_list()
+
+    l1.insert(3, 'u')
+
+    assert str(l1) == '[p, o, n, u, y]'
+
+def test_insert_fail():
+    l1 = create_list()
+
+    l1.insert(3, 'u')
+
+    assert str(l1) != '[p, o, n, y]'
+
+
+def test_remove_success():
+    l1 = create_list()
+    l1.remove(0)
+
+    l2 = create_list()
+    l2.remove(0)
+
+    assert l1 == l2
```

Comment:
If there were a bug in `remove` where it doesn't actually remove anything (for example the list stays the same), then this test wouldn't catch it.

You should compare with values that don't depend on the tested function. Best would be to compare with *literal* values like this:

```py
assert l1 == List.from_array(['p', 'o', 'n'])
```

Reply:
Here I wrote a test so that if the remove implementation is incorrect, the test will show that everything is correct. 

I was in doubt whether it was possible to take a specific list, (`['p', 'o', 'n', 'y']` for example) and test on it. It seemed to me that by changing the list in `create_list()`, all the tests would also have to pass validation.

And what would be more correct to write in assert? This:
```Python
assert l1 == List.from_array(['p', 'o', 'n'])
```
or this:
```Python
assert str(l1) == str(List.from_array(['p', 'o', 'n']))
```

Reply:
Of course, it would be nice if changing the list in `create_list()` wouldn't break the tests. However, that is unimportant, and very often not possible like in this case.

Our ultimate goal is to write tests, that fail if there are bugs automatically for us. One of the rules of writing such tests, is that the value that we compare with in `assert` statements (also called - expected value) should not depend on the function that we test.

To recap. Let's establish the terminology:

- `system under test` (sut) - is the term that defines the functionality being tested. 99% of the time it is just the function. I will use this term as the function name that the test is focused on testing.
- `actual`  - the value that `system_under_test` (sut) actually returns according to its real implementation
- `expected` -  the value we expect `sut(...)` to return

I recommend you to clearly identify all 3 of these for any test that you write

```py
actual = sut(...)
expected = ... 
assert actual == expected
```

You will find that each `assert` is composed of actual and expected values. The way how you create expected values should not use `sut(...)`, because we are not sure that function is fully correct **in the context of this test**, however if the function is not the system under test, we are free to use it. Fro example, `from_array` function is not the system under test in this test, so it's fine to use it.

As for how you would write `l1 == List.from_array(...)` or `str(l1) == str(List.from_array(...))`. Both should work fine. However, the first one is shorter, and more efficient (we don't need to create an intermediate string just for comparison of the lists). Also, recall that the `==` of the lists is not the system under test of this test, so we are free to use it in the `assert`.

In other words: explicitly identify a single system under test, and make sure that nothing else than the `actual` value depends on it.

Reply:
Did I understand you correctly? 

Expected value is what we expect to get as a result of the function and set it ourselves. 
If the function is written incorrectly, then we change the implementation of the function itself. Until the tests show: "✅".

But there is no way to do the opposite, change the expected value. The tests may pass validation, but the internal implementation of the function is incorrect.

Reply:
Yes the main idea is that the expected value always stays the same.